### ISSUE TITLE
Merge filter dependent on order of inputs/stage types?

### ISSUE THREAD

**Issue by chambbj** [PDAL/PDAL#1142]
@abellgithub Pertaining to our discussion last week...

I've pushed a branch (https://github.com/PDAL/PDAL/tree/merge-issues-with-tests) with some test cases that demonstrate some of the issues I'm seeing with merge -- namely, that when adding inputs to a merge filter, if the first is a reader, the second seems to be ignored.

Looks like it is independent of PipelineManager, and that adding multiple inputs to the writer may be a suitable workaround (at least for LAS).

**Comment by chambbj** [comment]
Just added a few more tests. Merging two filters (via merge filter or multiple inputs to writer) seems to behave the same regardless, but contains one point less than I would expect. If I filter and merge manually (via `pdal translate` -> `pdal merge`), I get the expected number of points.

**Comment by abellgithub** [comment]
These tests pass for me.  Do they not for you?  If not, what is the platform?  Other details?  Have you fixed something?

**Comment by chambbj** [comment]
Correct. This is failing for me on OS X. See below. Only difference from master should be the addition of the tests.

```
Bradleys-MacBook-Pro:build chambbj$ bin/pdal_filters_additional_merge_test
[==========] Running 12 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 12 tests from AdditionalMergeTest
[ RUN      ] AdditionalMergeTest.merge_filter_and_reader_with_manager
E
Done
[       OK ] AdditionalMergeTest.merge_filter_and_reader_with_manager (12 ms)
[ RUN      ] AdditionalMergeTest.merge_reader_and_filter_with_manager
E
Done
/Users/chambbj/loki/pdal/repo/test/unit/filters/AdditionalMergeTest.cpp:121: Failure
Value of: np
  Actual: 1065
Expected: 1420U
Which is: 1420
[  FAILED  ] AdditionalMergeTest.merge_reader_and_filter_with_manager (10 ms)
[ RUN      ] AdditionalMergeTest.merge_filter_and_reader_without_manager
E
Done
[       OK ] AdditionalMergeTest.merge_filter_and_reader_without_manager (9 ms)
[ RUN      ] AdditionalMergeTest.merge_reader_and_filter_without_manager
E
Done
/Users/chambbj/loki/pdal/repo/test/unit/filters/AdditionalMergeTest.cpp:205: Failure
Value of: np
  Actual: 1065
Expected: 1420U
Which is: 1420
[  FAILED  ] AdditionalMergeTest.merge_reader_and_filter_without_manager (9 ms)
[ RUN      ] AdditionalMergeTest.filter_and_reader_writer_inputs_with_manager
E
Done
[       OK ] AdditionalMergeTest.filter_and_reader_writer_inputs_with_manager (8 ms)
[ RUN      ] AdditionalMergeTest.reader_and_filter_writer_inputs_with_manager
E
Done
[       OK ] AdditionalMergeTest.reader_and_filter_writer_inputs_with_manager (9 ms)
[ RUN      ] AdditionalMergeTest.filter_and_reader_writer_inputs_without_manager
E
Done
[       OK ] AdditionalMergeTest.filter_and_reader_writer_inputs_without_manager (10 ms)
[ RUN      ] AdditionalMergeTest.reader_and_filter_writer_inputs_without_manager
E
Done
[       OK ] AdditionalMergeTest.reader_and_filter_writer_inputs_without_manager (9 ms)
[ RUN      ] AdditionalMergeTest.merge_two_filters_with_manager
E
Done
[       OK ] AdditionalMergeTest.merge_two_filters_with_manager (8 ms)
[ RUN      ] AdditionalMergeTest.merge_two_filters_without_manager
E
Done
[       OK ] AdditionalMergeTest.merge_two_filters_without_manager (8 ms)
[ RUN      ] AdditionalMergeTest.two_filters_writer_inputs_with_manager
E
Done
[       OK ] AdditionalMergeTest.two_filters_writer_inputs_with_manager (8 ms)
[ RUN      ] AdditionalMergeTest.two_filters_writer_inputs_without_manager
E
Done
[       OK ] AdditionalMergeTest.two_filters_writer_inputs_without_manager (8 ms)
[----------] 12 tests from AdditionalMergeTest (108 ms total)

[----------] Global test environment tear-down
[==========] 12 tests from 1 test case ran. (108 ms total)
[  PASSED  ] 10 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] AdditionalMergeTest.merge_reader_and_filter_with_manager
[  FAILED  ] AdditionalMergeTest.merge_reader_and_filter_without_manager

 2 FAILED TESTS
```

**Comment by abellgithub** [comment]
Yosemite?  El Capitan?

On Tue, Feb 23, 2016 at 2:27 AM, chambbj <notifications@github.com> wrote:

> Correct. This is failing for me on OS X. See below. Only difference from
> master should be the addition of the tests.
>
> Bradleys-MacBook-Pro:build chambbj$ bin/pdal_filters_additional_merge_test
> [==========] Running 12 tests from 1 test case.
> [----------] Global test environment set-up.
> [----------] 12 tests from AdditionalMergeTest
> [ RUN      ] AdditionalMergeTest.merge_filter_and_reader_with_manager
> E
> Done
> [       OK ] AdditionalMergeTest.merge_filter_and_reader_with_manager (12 ms)
> [ RUN      ] AdditionalMergeTest.merge_reader_and_filter_with_manager
> E
> Done
> /Users/chambbj/loki/pdal/repo/test/unit/filters/AdditionalMergeTest.cpp:121: Failure
> Value of: np
>   Actual: 1065
> Expected: 1420U
> Which is: 1420
> [  FAILED  ] AdditionalMergeTest.merge_reader_and_filter_with_manager (10 ms)
> [ RUN      ] AdditionalMergeTest.merge_filter_and_reader_without_manager
> E
> Done
> [       OK ] AdditionalMergeTest.merge_filter_and_reader_without_manager (9 ms)
> [ RUN      ] AdditionalMergeTest.merge_reader_and_filter_without_manager
> E
> Done
> /Users/chambbj/loki/pdal/repo/test/unit/filters/AdditionalMergeTest.cpp:205: Failure
> Value of: np
>   Actual: 1065
> Expected: 1420U
> Which is: 1420
> [  FAILED  ] AdditionalMergeTest.merge_reader_and_filter_without_manager (9 ms)
> [ RUN      ] AdditionalMergeTest.filter_and_reader_writer_inputs_with_manager
> E
> Done
> [       OK ] AdditionalMergeTest.filter_and_reader_writer_inputs_with_manager (8 ms)
> [ RUN      ] AdditionalMergeTest.reader_and_filter_writer_inputs_with_manager
> E
> Done
> [       OK ] AdditionalMergeTest.reader_and_filter_writer_inputs_with_manager (9 ms)
> [ RUN      ] AdditionalMergeTest.filter_and_reader_writer_inputs_without_manager
> E
> Done
> [       OK ] AdditionalMergeTest.filter_and_reader_writer_inputs_without_manager (10 ms)
> [ RUN      ] AdditionalMergeTest.reader_and_filter_writer_inputs_without_manager
> E
> Done
> [       OK ] AdditionalMergeTest.reader_and_filter_writer_inputs_without_manager (9 ms)
> [ RUN      ] AdditionalMergeTest.merge_two_filters_with_manager
> E
> Done
> [       OK ] AdditionalMergeTest.merge_two_filters_with_manager (8 ms)
> [ RUN      ] AdditionalMergeTest.merge_two_filters_without_manager
> E
> Done
> [       OK ] AdditionalMergeTest.merge_two_filters_without_manager (8 ms)
> [ RUN      ] AdditionalMergeTest.two_filters_writer_inputs_with_manager
> E
> Done
> [       OK ] AdditionalMergeTest.two_filters_writer_inputs_with_manager (8 ms)
> [ RUN      ] AdditionalMergeTest.two_filters_writer_inputs_without_manager
> E
> Done
> [       OK ] AdditionalMergeTest.two_filters_writer_inputs_without_manager (8 ms)
> [----------] 12 tests from AdditionalMergeTest (108 ms total)
>
> [----------] Global test environment tear-down
> [==========] 12 tests from 1 test case ran. (108 ms total)
> [  PASSED  ] 10 tests.
> [  FAILED  ] 2 tests, listed below:
> [  FAILED  ] AdditionalMergeTest.merge_reader_and_filter_with_manager
> [  FAILED  ] AdditionalMergeTest.merge_reader_and_filter_without_manager
>
>  2 FAILED TESTS
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PDAL/PDAL/issues/1142#issuecomment-187464928>.
>



-- 
Andrew Bell
andrew.bell.ia@gmail.com

**Comment by chambbj** [comment]
Yosemite

**Comment by chambbj** [comment]
Closed via #1143
